classes/village/VillageUpgradeManager.php

```diff
@@ -0,0 +1,368 @@
+            return false;
+        }
+        // check if the previous tier is built
+        if ($tier > 1 && $village->buildings[$building_key]->tier < $tier - 1) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * @param System $system
+     * @param Village $village
+     * @param string $building_key
+     * @return string
+     */
+    public static function beginConstruction(System $system, Village $village, $building_key): string {
+        // get construction costs and time
+        $materials_cost = VillageBuildingConfig::BUILDING_CONSTRUCTION_COST[$building_key][$village->buildings[$building_key]->tier + 1][WarManager::RESOURCE_MATERIALS];
+        $food_cost = VillageBuildingConfig::BUILDING_CONSTRUCTION_COST[$building_key][$village->buildings[$building_key]->tier + 1][WarManager::RESOURCE_FOOD];
+        $wealth_cost = VillageBuildingConfig::BUILDING_CONSTRUCTION_COST[$building_key][$village->buildings[$building_key]->tier + 1][WarManager::RESOURCE_WEALTH];
+        $progress_required = VillageBuildingConfig::BUILDING_CONSTRUCTION_TIME[$building_key][$village->buildings[$building_key]->tier + 1] * 86400;
+        // check if requirements met
+        if (!VillageUpgradeManager::checkConstructionRequirementsMet($village, $building_key, $village->buildings[$building_key]->tier + 1)) {
+            return "Construction requirements not met!";
+        }
+        // check if another building is already under construction
+        foreach ($village->buildings as $building) {
+            if ($building->status == VillageBuildingConfig::BUILDING_STATUS_UPGRADING) {
+                return "Another building is already under construction!";
+            }
+        }
+        // check if has previous progress
+        if (isset($village->buildings[$building_key]->construction_progress)) {
+            // check if the village has enough resources
+            if ($village->resources[WarManager::RESOURCE_MATERIALS] < $materials_cost || $village->resources[WarManager::RESOURCE_FOOD] < $food_cost || $village->resources[WarManager::RESOURCE_WEALTH] < $wealth_cost) {
+                return "Not enough resources!";
+            }
+            // update village resources
+            $village->subtractResource(WarManager::RESOURCE_MATERIALS, $materials_cost);
+            $village->subtractResource(WarManager::RESOURCE_FOOD, $food_cost);
+            $village->subtractResource(WarManager::RESOURCE_WEALTH, $wealth_cost);
+            $village->updateResources();
+            // log expenditure in resource_logs table
+            $system->db->query("INSERT INTO `resource_logs`(`village_id`, `resource_id`, `type`, `quantity`, `time`) VALUES ({$village->village_id}, " . WarManager::RESOURCE_MATERIALS . ", " . VillageManager::RESOURCE_LOG_CONSTRUCTION_COST . ", {$materials_cost}, " . time() . ")");
```

Comment:
Could/should we move this into `$village->subtractResource`? Is there a valid case for subtracting a resource without logging it?

Reply:
For the cron job we want the ability to get a query string, I wonder if we make subtract/addResource return a query string, or alternatively take an optional parameter that executes it automatically (updateResources works this way).